tests/test_astroid.py

```diff
@@ -1,15 +1,49 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals, print_function
+
+import re
+
 import astroid
 from . import tools, test_mark_tokens
 
 
 class TestAstroid(test_mark_tokens.TestMarkTokens):
 
   is_astroid_test = True
+  module = astroid
 
   @classmethod
   def create_mark_checker(cls, source):
     builder = astroid.builder.AstroidBuilder()
     tree = builder.string_build(source)
     return tools.MarkChecker(source, tree=tree)
+
+  def assert_nodes_equal(self, node1, node2):
+    self.assertEqual(
+      repr_tree(node1),
+      repr_tree(node2),
+    )
+
+
+def repr_tree(node):
+  """
+  Returns a canonical string representation of an astroid node
+  normalised to ignore the context of each node which can change when parsing
+  substrings of source code.
+
+  E.g. "a" is a Name in expression "a + 1" and is an AssignName in expression "a = 1",
+  but we don't care about this difference when comparing structure and content.
+  """
+  result = node.repr_tree()
+
+  # astroid represents context in multiple ways
+  # Convert Store and Del contexts to Load
+  # Similarly convert Assign/Del Name/Attr to just Name/Attribute (i.e. Load)
+  result = re.sub(r'(AssignName|DelName)(\(\s*name=)', r'Name\2', result)
+  result = re.sub(r'(AssignAttr|DelAttr)(\(\s*attrname=)', r'Attribute\2', result)
+  result = re.sub(r'ctx=<Context\.(Store: 2|Del: 3)>', r'ctx=<Context.Load: 1>', result)
+
+  # Weird bug in astroid that collapses spaces in docstrings sometimes maybe
+  result = re.sub(r"' +\\n'", r"'\\n'", result)
```

Comment:
This doesn't seem specific to docstrings. Is this really needed?

Reply:
It's really weird, without this everything happily works for like 50 files and then suddenly it trips on one file, and the diff between the strings looks like this:

```diff
*** n1.txt      2019-10-19 20:39:12.000000000 +0200
--- n2.txt      2019-10-19 20:39:12.000000000 +0200
***************
*** 1093,1101 ****
                  'value to stop\n'
                  '        them.\n'
                  '\n'
                  '        :param result: test result object\n'
!                 '        \n'
                  '        .. Note:: When tests are run under a test runner '
                  'other than\n'
                  '           :class:`nose.core.TextTestRunner`, such as\n'
                  '           via ``python setup.py test``, this method may be '
--- 1093,1101 ----
                  'value to stop\n'
                  '        them.\n'
                  '\n'
                  '        :param result: test result object\n'
!                 '\n'
                  '        .. Note:: When tests are run under a test runner '
                  'other than\n'
                  '           :class:`nose.core.TextTestRunner`, such as\n'
                  '           via ``python setup.py test``, this method may be '
***************
*** 1154,1162 ****
                  '        want to replace or modify the error tuple, return a '
                  'new error\n'
                  '        tuple, otherwise return err, the original error '
                  'tuple.\n'
!                 '        \n'
                  '        :param test: the test case\n'
                  '        :type test: :class:`nose.case.Test`\n'
                  '        :param err: sys.exc_info() tuple\n'
                  '        :type err: 3-tuple\n'
--- 1154,1162 ----
                  '        want to replace or modify the error tuple, return a '
                  'new error\n'
                  '        tuple, otherwise return err, the original error '
                  'tuple.\n'
!                 '\n'
                  '        :param test: the test case\n'
                  '        :type test: :class:`nose.case.Test`\n'
                  '        :param err: sys.exc_info() tuple\n'
                  '        :type err: 3-tuple\n'
***************
*** 1210,1218 ****
                  '        want to replace or modify the error tuple, return a '
                  'new error\n'
                  '        tuple, otherwise return err, the original error '
                  'tuple.\n'
!                 '        \n'
                  '        :param test: the test case\n'
                  '        :type test: :class:`nose.case.Test`\n'
                  '        :param err: sys.exc_info() tuple\n'
                  '        :type err: 3-tuple\n'
--- 1210,1218 ----
                  '        want to replace or modify the error tuple, return a '
                  'new error\n'
                  '        tuple, otherwise return err, the original error '
                  'tuple.\n'
!                 '\n'
                  '        :param test: the test case\n'
                  '        :type test: :class:`nose.case.Test`\n'
                  '        :param err: sys.exc_info() tuple\n'
                  '        :type err: 3-tuple\n'
***************
*** 2264,2272 ****
                 name='wantMethod',
                 doc=('Return true to collect this method as a test, false to\n'
                  '        prevent it from being collected, and None if you '
                  "don't care.\n"
!                 '        \n'
                  '        :param method: The method object being examined by '
                  'the selector\n'
                  '        :type method: unbound method\n'
                  '        '),
--- 2264,2272 ----
                 name='wantMethod',
                 doc=('Return true to collect this method as a test, false to\n'
                  '        prevent it from being collected, and None if you '
                  "don't care.\n"
!                 '\n'
                  '        :param method: The method object being examined by '
                  'the selector\n'
                  '        :type method: unbound method\n'
                  '        '),
```